core/src/main/java/kafka/server/NetworkPartitionMetadataClient.java

```diff
@@ -0,0 +1,305 @@
+        Map<String, ListOffsetsTopic> topicsMap = new HashMap<>();
+        partitions.forEach(tp -> {
+            if (!topicsMap.containsKey(tp.topic())) {
+                ListOffsetsTopic topic = new ListOffsetsTopic().setName(tp.topic());
+                topicsMap.put(tp.topic(), topic);
+            }
+            ListOffsetsTopic topic = topicsMap.get(tp.topic());
+            topic.partitions().add(
+                new ListOffsetsPartition()
+                    .setPartitionIndex(tp.partition())
+                    .setTimestamp(ListOffsetsRequest.LATEST_TIMESTAMP)
+                    .setCurrentLeaderEpoch(-1) // Will be set by broker if needed
+            );
+        });
+        // Isolation level will always be READ_UNCOMMITTED when finding the partition end offset.
+        return ListOffsetsRequest.Builder.forConsumer(
+            true,
+            IsolationLevel.READ_UNCOMMITTED
+        ).setTargetTimes(List.copyOf(topicsMap.values()));
+    }
+
+    /**
+     * Handles the response from a ListOffsets request.
+     */
+    // Visible for Testing.
+    void handleResponse(Map<TopicPartition, CompletableFuture<OffsetResponse>> partitionFutures, ClientResponse clientResponse) {
+        // Handle error responses first
+        if (handleErrorResponse(partitionFutures, clientResponse)) {
+            return;
+        }
+
+        log.debug("ListOffsets response received successfully - {}", clientResponse);
+        ListOffsetsResponse response = (ListOffsetsResponse) clientResponse.responseBody();
+
+        for (ListOffsetsTopicResponse topicResponse : response.topics()) {
+            String topicName = topicResponse.name();
+            for (ListOffsetsPartitionResponse partitionResponse : topicResponse.partitions()) {
+                TopicPartition tp = new TopicPartition(topicName, partitionResponse.partitionIndex());
+                // Get the corresponding future from the map and complete it.
+                CompletableFuture<OffsetResponse> future = partitionFutures.get(tp);
+                if (future != null) {
+                    future.complete(new OffsetResponse(partitionResponse.offset(), Errors.forCode(partitionResponse.errorCode())));
+                }
+            }
+        }
+
+        partitionFutures.forEach((tp, future) -> {
+            // If future is not completed yet hence topic-partition was not included in the response, complete with error
+            if (!future.isDone()) {
+                future.complete(new OffsetResponse(-1, Errors.UNKNOWN_TOPIC_OR_PARTITION));
+            }
+        });
+    }
+
+    /**
+     * Handles error responses by completing all associated futures with an error. Returns true if an error was
+     * handled. Otherwise, returns false.
+     */
+    private boolean handleErrorResponse(Map<TopicPartition, CompletableFuture<OffsetResponse>> partitionFutures, ClientResponse clientResponse) {
```

Comment:
Not neccessarily there could be error.
```suggestion
    private boolean maybeHandleErrorResponse(Map<TopicPartition, CompletableFuture<OffsetResponse>> partitionFutures, ClientResponse clientResponse) {
```